Source/Core/Runtime/Timers/MachineTimer.cs

```diff
@@ -0,0 +1,156 @@
+    internal sealed class MachineTimer : IMachineTimer
+    {
+        /// <summary>
+        /// The unique name of this timer.
+        /// </summary>
+        public string Name { get; private set; }
+
+        /// <summary>
+        /// The machine that owns this timer.
+        /// </summary>
+        private readonly Machine Owner;
+
+        /// <summary>
+        /// The id of the machine that owns this timer.
+        /// </summary>
+        public MachineId OwnerId => this.Owner.Id;
+
+        /// <summary>
+        /// The internal timer.
+        /// </summary>
+        private Timer InternalTimer;
+
+        /// <summary>
+        /// The timeout event.
+        /// </summary>
+        private readonly TimerElapsedEvent TimeoutEvent;
+
+        /// <summary>
+        /// The amount of time to wait before sending the first timeout event.
+        /// </summary>
+        public TimeSpan DueTime { get; private set; }
+
+        /// <summary>
+        /// The time interval between timeout events.
+        /// </summary>
+        public TimeSpan Period { get; private set; }
+
+        /// <summary>
+        /// The optional payload of the timer. This is null if there is no payload.
+        /// </summary>
+        public object Payload { get; private set; }
+
+        /// <summary>
+        /// Creates a new instance of the <see cref="MachineTimer"/> class.
+        /// </summary>
+        /// <param name="owner">The machine that owns this timer.</param>
+        /// <param name="dueTime">The amount of time to wait before sending the first timeout event.</param>
+        /// <param name="period">The time interval between timeout events.</param>
+        /// <param name="name">Optional unique name of the timer.</param>
+        /// <param name="payload">Optional payload of the timeout event.</param>
+        public MachineTimer(Machine owner, TimeSpan dueTime, TimeSpan period, string name, object payload)
+        {
+            this.Name = name ?? Guid.NewGuid().ToString();
+            this.Owner = owner;
+            this.DueTime = dueTime;
+            this.Period = period;
+            this.Payload = payload;
+
+            this.Owner.Runtime.Logger.OnCreateTimer(this, this.Owner.Id);
+
+            this.TimeoutEvent = new TimerElapsedEvent(this);
+            this.InternalTimer = new Timer(HandleTimeout, null, this.DueTime, Timeout.InfiniteTimeSpan);
+        }
+
+        /// <summary>
+        /// Handles the timeout.
+        /// </summary>
+        private void HandleTimeout(object state)
+        {
+            // Send a timeout event.
+            this.Owner.Runtime.SendEvent(this.Owner.Id, this.TimeoutEvent);
+
+            if (this.Period.Milliseconds >= 0)
+            {
+                // The timer is periodic, so schedule the next timeout.
+                try
+                {
+                    // Start the next timeout period.
+                    this.InternalTimer.Change(this.Period, Timeout.InfiniteTimeSpan);
```

Comment:
Check if return value was `false`? The documentation of `Change` says:

        // Returns:
        //     true if the timer was successfully updated; otherwise, false.

It can also throw exceptions for illegal arguments. Perhaps assert for them at the beginning of `StartTimer` so that errors are not dropped silently?

Reply:
These are already asserted at the beginning of `StartTimer` and `StartPeriodicTimer`. It seems that `Change` will "never" return `false`, where never is more like extremely unlikely, as discussed [here](https://stackoverflow.com/questions/12586932/does-timer-change-ever-return-false).